vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/PriorityByWorkflow.java

```diff
@@ -117,29 +93,38 @@ public synchronized ConsumableResourceResponse request(
               + "values should be one of the following: %s", workflowName, workflowPriority,
               acceptedPriorities.stream().map(String::valueOf).collect(Collectors.joining(", "))));
     }
+
+
     final var state = workflows.get(workflowName);
-    if (state == null) {
-      return ConsumableResourceResponse.error(
-          String.format("Internal Vidarr error: The %s workflow run priority has not been configured properly.", workflowName));
+    if (state == null || state.waiting.isEmpty()) {
+      return ConsumableResourceResponse.AVAILABLE;
+    }
+
+    // If this workflow has already been seen
+    // Add the current run to the waitlist
+    // ensuring it replaces previous runs with the same ID, accounting for if the priority has changed
+    set(workflowName, vidarrId, input);
+    SimpleEntry resourcePair = new SimpleEntry(vidarrId, workflowPriority);
+
+    if (workflowPriority >= state.waiting.last().getValue()) {
```

Comment:
Thought about this more over the weekend; why would the waiting list ever populate?
State.waiting will be empty on startup, right? So the test at 99 will always pass, and the waiting list will continue to be empty. If a run is stopped by max-in-flight or another consumable resource before or after checking priority, it can't get added to this waiting list because all the consumable resources are isolated from each other.

Reply:
I think that could be the case if `set() ` was only ever called in `request()` but it's also called [here](https://github.com/oicr-gsi/vidarr/blob/add_required_reource_check/vidarr-server/src/main/java/ca/on/oicr/gsi/vidarr/server/Main.java#L765), which will populate the waiting list.

Reply:
In addition to being populated on startup, this PR also calls `set` during `release`, so a workflow run may get added to state.waiting if priority gets checked and then the workflow run gets throttled by max-in-flight. (It's not a perfect system because as you say, the consumable resources are isolated from each other and there are no guarantees about order of CR checking)

Reply:
I _had_ planned to remove that call from `release` based on the conversation from Friday, since having that will add it to the wait list both when it's throttled by something else (ideally) but also when the workflow succeeds.

Reply:
Yeah, counting up every succeeded workflow run in the waiting state will effectively create a memory leak

Reply:
Here's my thoughts:
1. `where(ACTIVE_WORKFLOW_RUN.ENGINE_PHASE.eq(Phase.WAITING_FOR_RESOURCES))` can potentially be removed from the start-up. However, that still doesn't resolve the issue if a new workflow is added.
2.Modify release to have a success or failure "flag" `public void release(String workflowName, String workflowVersion, String vidarrId, boolean throttled) ` where
```
if throttled:
	add to wait list using set(default priority)
if not throttled:
	do nothing
```

Reply:
This PriorityByWorkflow feature is most useful when a new workflow launches dozens of other workflows. In that case, a majority of the workflow runs will be throttled by max-in-flight. The way consumable resources get checked, the first set of workflow runs that get checked will launch (likely irrespective of priority); it's the next round of checking, when max-in-flight throttling is in place, that launching by priority will be in effect until the bolus of workflow runs completes.